Improving dev documentation for inline help on form. (#4326)
jenkins-infra/jenkins.io: content/doc/developer/forms/jelly-form-controls.adoc

  title: ui-samples-plugin
---

Most of the the jelly files in the Jenkins source have embedded documentation. See link:/doc/developer/development-environment/ide-configuration[IDE Configuration] for details on how to get this setup so you can click through to it in your IDE. 

> For example you can see the embedded documentation of link:https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/lib/form/form.jelly[ the <form> tag ].

Reviewer: Beware that floating branch links such as this can become invalid after refactorings.

Author: yes I don't know how to do it the best way, but I think the doc should link to jelly content as the documentation is live there.

Author: Changed

   <f:select />
</f:entry>
----

== Inline help

* jelly definition link:https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/lib/form/helpLink.jelly[form/helpLink.jelly]

> Most of form elements allows to have an inline help generated by adding a small html file and following conventions.

Reviewer: ```suggestion
Most form elements allow you to have inline help generated by adding a small HTML file and following some conventions.
```

> For example <f:entry> tag, uses the field name to find the

Reviewer: ```suggestion
For example the `<f:entry>` tag uses the `field` attribute.
```

(not sure about exotic AsciiDoc syntax in the first case, may need `+` or `++` or something)

Given a file in src/main/resources/myForm.jelly with this content:

[source, xml]
----

> <f:entry title="Visible" field="isVisible">

Reviewer: Not the best example since conventionally the field would be called `visible` corresponding to a JavaBeans getter/setter `setVisible` / `isVisible`.

Author: I copied my morning example from https://github.com/jenkinsci/promoted-builds-plugin/blob/master/src/main/resources/hudson/plugins/promoted_builds/PromotionProcess/help-isVisible.html

I can change it to something better, like name

Author: changed

   <f:select />
</f:entry>
----

== Inline help

* jelly definition link:https://github.com/jenkinsci/jenkins/blob/63f80114e99f6692812c3039407652592bdf36fe/core/src/main/resources/lib/form/helpLink.jelly[form/helpLink.jelly]

=== Add an inline help to a form element

Most form elements allow you to have inline help generated by adding a small HTML file and following some conventions.

For example the `<f:entry>` tag uses the `field` attribute.

Given a file in src/main/resources/myForm.jelly with this content:

[source, xml]
----
<f:entry title="Name" field="name">
    <f:textbox />
</f:entry>
----

Adding a src/main/resources/help-name.html:
[source, html]
----
<div>
This is my content to help the end user understanding how to use this field.
</div>
----

Will automatically display the help button with the `<div>...</div>` content.

=== Additional notes on inline help

* Most controls support `help.html` as overall help for the `Describable`.

> * On some corner case, the help message can be hard coded in jelly with `help=` attribute, but please use build-in help as much as possible.

Reviewer 2: I don't think this is worded the best, possibly this is better?

```suggestion
* The help message can be overridden in jelly with the `help` attribute, but please use the convention 'help-fieldName.html' as much as possible.
```
